### PR TITLE
Created a new API key product type

### PR DESCRIPTION
**Pull request by BelleNottelling**
This PR introduces a new module named "Serviceapikey". It's purpose is to allow FOSSBilling administrators to sell API keys that can then be used in their application.
The module features the following API endpoints:

1. `/admin/serviceapikey/update` - Used to allow administrators to update the config and validity of an API key.
2. `/admin/serviceapikey/reset` - Used to allow administrators to reset and regenerate an API key.
4. `/client/serviceapikey/reset` - Allows a client to reset their API key.
5. `/guest/serviceapikey/check` - Checks if an API key is valid or not.
6. `/guest/serviceapikey/get_info` - Similar to the `check` endpoint, except that this will also return any custom params that are defined for the API key. This means you can set custom params such as a monthly request limit and your application can then pull that data from FOSSBilling + if it's valid or not.

In addition to having the ability define a specific API key length, administrators can also add custom parameters for a product / order:
Animation

When getting the info for an API key, these custom params will be included in a response like this:
```JSON
{
  "result": {
    "valid": 1,
    "id": 8,
    "config": { "monthlyLimit": 250, "somethingElse": 1150, "astring": "words" }
  },
  "error": null
}
```

These custom parameters allow the API key product to be fairly flexible, while maintaining simplicity. I can't think of any situations where it wouldn't be sufficient for usage with a custom application.

### CODE REVIEW

File: src/library/Model/ServiceApiKey.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * Copyright 2022-2023 FOSSBilling
+ * Copyright 2011-2021 BoxBilling, Inc.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * @copyright FOSSBilling (https://www.fossbilling.org)
+ * @license http://www.apache.org/licenses/LICENSE-2.0 Apache-2.0
+ */
+
+class Model_ServiceApiKey extends \RedBeanPHP\SimpleModel
```
Comment (BelleNottelling) [inline src/library/Model/ServiceApiKey.php:11]: Something to note:
Because this module doesn't use the deprecated logic that the existing "service" module's do, it doesn't actually seem like this model is used at all. If/when we migrate to a different ORM, this will still be useful, but even if we don't it also means that anyone can go and look at what the properties are and their types without needing to look at the SQL structure for it.

**Comment by BelleNottelling** [comment]
Okay. I just tested the entropy of the key generator (in the most inefficient way possible), and I think we are good to go on that front.
I made PHP generate 250 million keys, store them in RAM, and then check to see if any were duplicated.
- At the default 32 char length for the keys, 0 were duplicated and PHP managed to consume 37.9 GiB of RAM.
- With a 16 char length for the keys, 0 were duplicated and PHP managed to consume 34.18 GiB of RAM.
- And finally, with an 8 char length, 6,997,314 were duplicated and PHP managed to consume 32.32 GiB of RAM.

TLDR:
Even wth 250 million keys generated and **half** of the default API key length none of them were duplicated.
Also, if you force PHP to store 250 million values in an array and then compare them to try and find duplicates, you can cause it to use over 32GB of RAM.

Also for anyone who was curious, I did the same test with the [Servicelicense](https://github.com/FOSSBilling/FOSSBilling/blob/main/src/modules/Servicelicense/Plugin/Simple.php#L32) generator function.
- Length of 32: 0 duplicated, 36.09 GiB
- Length of 16: 0 duplicated, 32.37 GiB
- Length of 8: 15,531,359 duplicated, 30.47 GiB

Oh yeah, and across the board it was many times slower compared to the code I created for this product type.
So it is much slower and about twice as likely to produce a collision. In this hugely inefficient test, it did use less RAM, but in a more reasonable test of 100 keys both tests used 2mb of ram.

**Comment by BelleNottelling** [comment]
Okay. I think this is ready for it's initial round of reviews. Feel free to point out any potential issues or things that seem odd (and please do)
I've also created the rough draft for the docs for this product type, which should make it pretty straightforward for anyone to use it: https://fossbilling-o7m4npgls-fossbilling.vercel.app/docs/product-types/apikeys

**Comment by BelleNottelling** [comment]
The following changes has been done compared to the initial PR description:

## More configuration options
Products can have the capitalization, dashes, and length configured per the administrator's preference. The defaults are shown below:
image

## Slightly simplified logic
The original version of this used the API key's DB ID for API interactions and also gave an API key a `valid` column.
This has been changed to match the behavior of other modules where the API uses the order ID & uses the order's status to determine if it's a valid key or not,

## Tweaked API Output
 - Previously all custom parameters would be returned as a string, however now it will correctly detect if a custom parameter is numeric and will then return the value as a float instead of as a string
 - The `id` is no longer returned, as it only has use for internal FOSSBilling logic

### Misc
 - The email templates now are HTML based and match our new ones
 - I've updated the docs for it to reflect these changes and correct a few minor issues. New version is [here](https://fossbilling-faw7r1gjk-fossbilling.vercel.app/docs/product-types/apikeys)
 - It has an icon now and looks like this:
image

I believe this product type is now fully implemented & ready to be deployed

**Review by jaapmarcus — APPROVED** [review]

**Review by evrifaessa — APPROVED** [review]
Looks great :)
